### PR TITLE
Attempts to bring Dynamic more up to date

### PR DESCRIPTION
**Pull request by Crossedfall**
## About The Pull Request

Our version of dynamic is shamefully behind... This attempts to resolve that. Ports the following:

- https://github.com/tgstation/tgstation/pull/50013
- https://github.com/tgstation/tgstation/pull/51315
- https://github.com/tgstation/tgstation/pull/55176
- https://github.com/tgstation/tgstation/pull/55239
- https://github.com/tgstation/tgstation/pull/55200
- https://github.com/tgstation/tgstation/pull/56122
- https://github.com/tgstation/tgstation/pull/56252
- https://github.com/tgstation/tgstation/pull/56221 -- Big
- https://github.com/tgstation/tgstation/pull/56778
- https://github.com/tgstation/tgstation/pull/56961
- https://github.com/tgstation/tgstation/pull/57036
- https://github.com/tgstation/tgstation/pull/57093
- https://github.com/tgstation/tgstation/pull/57175 -- Big
- https://github.com/tgstation/tgstation/pull/57248

## Why It's Good For The Game
Dynamic is a great game mode, it's just held back by a lack of any updates.

## Changelog
:cl:TheChosenEvilOne, Rohesie, IndieanaJones, Mothblocks, Qustinnus
del: Removed high population override from dynamic.
fix: Fixed abductors being listed at round end on Dynamic despite none existing
fix: Dynamic abductors should no longer be split up between two different ships.
fix: Dynamic can now be configured when made the forced secret mode.
tweak: dynamic threat reports are no longer perfectly accurate, they diverge and are sometimes downright wrong.
add: Dynamic now rolls a separate budget for round start and midround antagonists, meaning less round start antags will roll in favor of more midround ones.
admin: The antag_cap value now uses an equation that scales with population, rather than a fixed array.
admin: autotraitor_cooldown now uses deciseconds rather than ticks, so the older value of 450 now means 45 seconds. Set it to be 9000 if you want it to be accurate to the old value (15 minutes), or simply unset it.
admin: Midround/latejoin dynamic timers are now configurable.
admin: Syndicate Sleeper Agent will now give details when ti fails.
fix: The autotraitor cooldown now respects configuration, rather than resetting back to 15 minutes after it is rolled.
fix: Fixed dynamic parameters not checking the configuration. This means that Manuel will now have slower midrounds, as was intended.
add: you can now cast rituals on dynamic if the threat level is 100
fix: Fixed multiple high impact rulesets being chosen on Dynamic.
fix: Midround dynamic injection no longer has a 100% chance of activating.
add: Dynamic will now play a part in controlling random event antagonists. If a dynamic midround injection is coming too soon, or too early, then the next midround injection will be buffed. Otherwise, it'll spawn normally.
admin: Admins can now cancel or replace midround rulesets.
fix: Admins can now correctly force midround rulesets.
/:cl:

### CODE REVIEW

**Comment by Sinestia** [comment]
Yay, 2021 is here

**Comment by Crossedfall** [comment]
On hold, pending ~~#3798~~ ~~#3803~~

**Comment by Crossedfall** [comment]
Note: I did not port https://github.com/tgstation/tgstation/issues/55956

**Comment by Crossedfall** [comment]
Failed my test merge.. Works locally, so I'll need to do some added troubleshooting.

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 1]
```diff
@@ -537,14 +536,14 @@
...
+	flags = LONE_RULESET
 	requirements = list(101,101,101,101,101,101,101,101,101,101)
-	high_population_requirement = 101
 	antag_cap = list(1,1,1,2,2,2,3,3,3,4)
 
-/datum/dynamic_ruleset/roundstart/devil/pre_execute()
-	var/num_devils = antag_cap[indice_pop]
-	mode.antags_rolled += num_devils
-
+/datum/dynamic_ruleset/roundstart/traitor/pre_execute(population)
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm]:
```suggestion
/datum/dynamic_ruleset/roundstart/devil/pre_execute(population)
```

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 2]
```diff
@@ -591,15 +590,15 @@
...
+	flags = LONE_RULESET
 	var/players_per_carrier = 30
 	var/monkeys_to_win = 1
 	var/escaped_monkeys = 0
 	var/datum/team/monkey/monkey_team
 
-/datum/dynamic_ruleset/roundstart/monkey/pre_execute()
-	var/carriers_to_make = max(round(mode.roundstart_pop_ready / players_per_carrier, 1), 1)
-	mode.antags_rolled += carriers_to_make
+/datum/dynamic_ruleset/roundstart/traitor/pre_execute(population)
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm]:
```suggestion
/datum/dynamic_ruleset/roundstart/monkey/pre_execute(population)
```

**Review by park66665 — COMMENTED** [review]
This might've been responsible for the problems at the testmerges

**Comment by Crossedfall** [comment]
Local tests and TM was solid. No soft-locks.

File: code/game/gamemodes/dynamic/ruleset_picking.dm [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	if (rule.execute())
+		log_game("DYNAMIC: Injected a [rule.ruletype == "latejoin" ? "latejoin" : "midround"] ruleset [rule.name].")
+		if(rule.flags & HIGH_IMPACT_RULESET)
+			high_impact_ruleset_executed = TRUE
+		else if(rule.flags & ONLY_RULESET)
+			only_ruleset_executed = TRUE
+		if(rule.ruletype == "Latejoin")
+			var/mob/M = pick(rule.candidates)
+			message_admins("[key_name(M)] joined the station, and was selected by the [rule.name] ruleset.")
+			log_game("DYNAMIC: [key_name(M)] joined the station, and was selected by the [rule.name] ruleset.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/ruleset_picking.dm]:
```suggestion
			dynamic_log("[key_name(M)] joined the station, and was selected by the [rule.name] ruleset.")
```

File: code/game/gamemodes/dynamic/ruleset_picking.dm [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	return FALSE
+
+/// Fired when an admin cancels the current midround injection.
+/datum/game_mode/dynamic/proc/admin_cancel_midround(mob/user, timer_id)
+	if (midround_injection_timer_id != timer_id || !deltimer(midround_injection_timer_id))
+		to_chat(user, "<span class='notice'>Too late!</span>")
+		return
+
+	log_admin("[key_name(user)] cancelled the next midround injection.")
+	message_admins("[key_name(user)] cancelled the next midround injection.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/ruleset_picking.dm]:
```suggestion
	dynamic_log("[key_name(user)] cancelled the next midround injection.")
```

File: code/game/gamemodes/dynamic/ruleset_picking.dm [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+/datum/game_mode/dynamic/proc/admin_different_midround(mob/user, timer_id)
+	if (midround_injection_timer_id != timer_id || !deltimer(midround_injection_timer_id))
+		to_chat(user, "<span class='notice'>Too late!</span>")
+		return
+
+	midround_injection_timer_id = null
+
+	if (isnull(current_midround_rulesets) || current_midround_rulesets.len == 0)
+		log_admin("[key_name(user)] asked for a different midround injection, but there were none left.")
+		message_admins("[key_name(user)] asked for a different midround injection, but there were none left.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/ruleset_picking.dm]:
```suggestion
		dynamic_log("[key_name(user)] asked for a different midround injection, but there were none left.")
```

File: code/game/gamemodes/dynamic/ruleset_picking.dm [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		return
+
+	log_admin("[key_name(user)] asked for a different midround injection.")
+	message_admins("[key_name(user)] asked for a different midround injection.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/ruleset_picking.dm]:
```suggestion
	dynamic_log("[key_name(user)] asked for a different midround injection.")
```

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 7]
```diff
@@ -179,26 +216,50 @@ GLOBAL_VAR_INIT(dynamic_forced_threat_level, -1)
 		show_threatlog(usr)
 	else if (href_list["stacking_limit"])
 		GLOB.dynamic_stacking_limit = input(usr,"Change the threat limit at which round-endings rulesets will start to stack.", "Change stacking limit", null) as num
+	else if(href_list["force_latejoin_rule"])
+		var/added_rule = input(usr,"What ruleset do you want to force upon the next latejoiner? This will bypass threat level and population restrictions.", "Rigging Latejoin", null) as null|anything in sortList(latejoin_rules)
+		if (!added_rule)
+			return
+		forced_latejoin_rule = added_rule
+		log_admin("[key_name(usr)] set [added_rule] to proc on the next latejoin.")
+		message_admins("[key_name(usr)] set [added_rule] to proc on the next latejoin.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic.dm]:
```suggestion
		dynamic_log("[key_name(usr)] set [added_rule] to proc on the next latejoin.")
```

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 8]
```diff
@@ -179,26 +216,50 @@ GLOBAL_VAR_INIT(dynamic_forced_threat_level, -1)
...
+	else if(href_list["clear_forced_latejoin"])
+		forced_latejoin_rule = null
+		log_admin("[key_name(usr)] cleared the forced latejoin ruleset.")
+		message_admins("[key_name(usr)] cleared the forced latejoin ruleset.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic.dm]:
```suggestion
		dynamic_log("[key_name(usr)] cleared the forced latejoin ruleset.")
```

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 9]
```diff
@@ -179,26 +216,50 @@ GLOBAL_VAR_INIT(dynamic_forced_threat_level, -1)
...
+	else if(href_list["force_midround_rule"])
+		var/added_rule = input(usr,"What ruleset do you want to force right now? This will bypass threat level and population restrictions.", "Execute Ruleset", null) as null|anything in sortList(midround_rules)
+		if (!added_rule)
+			return
+		log_admin("[key_name(usr)] executed the [added_rule] ruleset.")
+		message_admins("[key_name(usr)] executed the [added_rule] ruleset.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic.dm]:
```suggestion
		dynamic_log("[key_name(usr)] executed the [added_rule] ruleset.")
```

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 10]
```diff
@@ -537,14 +536,14 @@
 	required_candidates = 1
 	weight = 3
 	cost = 0
+	flags = LONE_RULESET
 	requirements = list(101,101,101,101,101,101,101,101,101,101)
-	high_population_requirement = 101
 	antag_cap = list(1,1,1,2,2,2,3,3,3,4)
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm]:
```suggestion
	antag_cap = list("denominator" = 30)
```

We need to give it an update.

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 11]
```diff
@@ -627,13 +604,16 @@ GLOBAL_VAR_INIT(dynamic_forced_threat_level, -1)
 		message_admins("DYNAMIC: Checking for midround injection.")
 		log_game("DYNAMIC: Checking for midround injection.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic.dm]:
```suggestion
		dynamic_log("Checking for midround injection.")
```

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 12]
```diff
@@ -263,40 +321,46 @@ GLOBAL_VAR_INIT(dynamic_forced_threat_level, -1)
...
+	log_game("DYNAMIC: Centre is [threat_curve_centre], Width is [threat_curve_width], Forced extended is [GLOB.dynamic_forced_extended ? "Enabled" : "Disabled"], No stacking is [GLOB.dynamic_no_stacking ? "Enabled" : "Disabled"].")
+	log_game("DYNAMIC: Stacking limit is [GLOB.dynamic_stacking_limit].")
 	if(GLOB.dynamic_forced_threat_level >= 0)
 		threat_level = round(GLOB.dynamic_forced_threat_level, 0.1)
-		threat = threat_level
 	else
 		generate_threat()
+	generate_budgets()
+	set_cooldowns()
+	log_game("DYNAMIC: Dynamic Mode initialized with a Threat Level of... [threat_level]! ([round_start_budget] round start budget)")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic.dm]:
```suggestion
	dynamic_log("Dynamic Mode initialized with a Threat Level of... [threat_level]! ([round_start_budget] round start budget)")
```

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 13]
```diff
@@ -308,40 +372,49 @@ GLOBAL_VAR_INIT(dynamic_forced_threat_level, -1)
...
 
 	if(GLOB.dynamic_forced_roundstart_ruleset.len > 0)
 		rigged_roundstart()
+	else if(valid_roundstart_ruleset < 1)
+		log_game("DYNAMIC: [valid_roundstart_ruleset] enabled roundstart rulesets.")
+		return TRUE
 	else
 		roundstart()
 
+	log_game("DYNAMIC: [round_start_budget] round start budget was left, donating it to midrounds.")
```
Comment (park66665) [inline code/game/gamemodes/dynamic/dynamic.dm]:
```suggestion
	dynamic_log("[round_start_budget] round start budget was left, donating it to midrounds.")
```

**Review by park66665 — COMMENTED** [review]
The code looks good (actually I am a bit lost in the expansiveness of the code)

**Review by park66665 — APPROVED** [review]
Let's see how things turn out. YOLO?

**Pull request merged**